include/rmm/mr/device/detail/arena.hpp

```diff
@@ -0,0 +1,616 @@
+  {
+    RMM_LOGGING_ASSERT(size >= sz);
+    if (size > sz) {
+      return {{pointer, sz, is_head}, {increment(pointer, sz), size - sz, false}};
+    } else {
+      return {*this, {}};
+    }
+  }
+
+  /**
+   * @brief Coalesce two contiguous blocks into one.
+   *
+   * `this` must immediately precede `b` and both `this` and `b` must be from the same superblock.
+   *
+   * @param b block to merge.
+   * @return block The merged block.
+   */
+  block merge(block const& b) const
+  {
+    RMM_LOGGING_ASSERT(is_contiguous_before(b));
+    return {pointer, size + b.size, is_head};
+  }
+
+  /**
+   * @brief Increment the given pointer by the given size.
+   *
+   * @param pointer the pointer to increment.
+   * @param size the size to add to the pointer.
+   * @return void* The resulting pointer.
+   */
+  static void* increment(void* pointer, std::size_t size)
+  {
+    return static_cast<void*>(static_cast<char*>(pointer) + size);
+  }
+
+  /// Used by std::set to compare blocks.
+  bool operator<(block const& b) const { return pointer < b.pointer; }
+};
+
+/// The required allocation alignment.
+constexpr std::size_t allocation_alignment = 256;
+
+/**
+ * @brief Align up to the allocation alignment.
+ *
+ * @param[in] v value to align
+ * @return Return the aligned value
+ */
+constexpr std::size_t align_up(std::size_t v) noexcept
+{
+  return rmm::detail::align_up(v, allocation_alignment);
+}
+
+/**
+ * @brief Align down to the allocation alignment.
+ *
+ * @param[in] v value to align
+ * @return Return the aligned value
+ */
+constexpr std::size_t align_down(std::size_t v) noexcept
+{
+  return rmm::detail::align_down(v, allocation_alignment);
+}
+
+/**
+ * @brief Get the first free block of at least `size` bytes.
+ *
+ * Address-ordered first-fit has shown to perform slightly better than best-fit when it comes to
+ * memory fragmentation, and slightly cheaper to implement. It is also used by some popular
+ * allocators such as jemalloc.
+ *
+ * \see Johnstone, M. S., & Wilson, P. R. (1998). The memory fragmentation problem: Solved?. ACM
+ * Sigplan Notices, 34(3), 26-36.
+ *
+ * @param free_blocks The address-ordered set of free blocks.
+ * @param size The number of bytes to allocate.
+ * @return block A block of memory of at least `size` bytes, or an empty block if not found.
+ */
+inline block first_fit(std::set<block>& free_blocks, std::size_t size)
+{
+  auto const iter = std::find_if(
+    free_blocks.cbegin(), free_blocks.cend(), [size](auto const& b) { return b.fits(size); });
+
+  if (iter == free_blocks.cend()) {
+    return {};
+  } else {
+    // Remove the block from the free_list.
+    auto const b = *iter;
+    auto const i = free_blocks.erase(iter);
+
+    // Split the block and put the remainder back.
+    auto const split     = b.split(size);
+    auto const remainder = split.second;
+    if (remainder.is_valid()) { free_blocks.insert(i, remainder); }
+    return split.first;
+  }
```

Comment:
This looks like the same logic as `split_block`. Can that be called instead?

Reply:
It's slightly different. Since we have the iterator, we can insert directly in front of the iterator in O(1). In `split_block` we have to do a search that's O(log(n)). Maybe this is too much premature optimization. :)